### PR TITLE
Add overlap with DDP to ZeRO

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#62052 Add overlap with DDP to ZeRO**

Differential Revision: [D29895933](https://our.internmc.facebook.com/intern/diff/D29895933)

### CODE REVIEW

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -856,6 +857,7 @@ def forward(self, *inputs, **kwargs):
             # This should be called only once during whole training period.
             if torch.is_grad_enabled() and self.reducer._rebuild_buckets():
                 logging.info("Reducer buckets have been rebuilt in this iteration.")
+                self._has_rebuilt_buckets = True
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:860]: Note that this might not be entirely accurate. 

If find_unused_params=True then we don't actually rebuilt buckets and it's basically a noop (you can confirm by seeing reducer code). Not sure if it really affects your use case though.

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+        def zero_step(_) -> List[torch.Tensor]:
+            r"""
+            Performs a partial :class:`ZeroRedundancyOptimizer` :meth:`step`
+            using the gradients in the given :class:`DistributedDataParallel`
+            gradient bucket.
+            """
+            # Proceed as normal until the DDP buckets have been rebuilt
+            if not ddp._has_rebuilt_buckets:
+                return [bucket.get_tensor()]
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:58]: This means we don't do optimizer step at all before buckets are rebuilt?

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                params_per_bucket.append(bucket_params)
+
+                return [bucket.get_tensor()]
+
+            if rank_to_update == rank:
+                assert len(zero.optim.param_groups) == 1, \
+                    "Overlapping DDP with ZeRO only supports a single " \
+                    "parameter group"
+                # Construct the `gradients` input for the local optimizer step,
+                # which expects `None` in a list position to indicate that the
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:87]: That might be a little bit brittle or easy for future developer to forget, can you do something like have a global:

_ZERO_NO_UPDATE_PARAM = None 

And use that constant instead, might make the intent more clear.

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            if overlap_info.status == _OverlapStatus.UNINITIALIZED:
+                overlap_info.status = _OverlapStatus.DDP_HAS_REBUILT_BUCKETS
+
+            bucket_params = bucket.get_model_params_for_bucket()
+            assert len(bucket_params) > 0, "Empty bucket"
+            rank_to_update = zero._ddp_bucket_index_to_rank(bucket_index)
+
+            # Once DDP buckets have been rebuilt but ZeRO has not been
+            # properly initialized yet, collect the information needed
+            if overlap_info.status == _OverlapStatus.DDP_HAS_REBUILT_BUCKETS:
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:72]: I don't see a state transition to initialized, where does that happen?

**Review by rohan-varma — COMMENTED** [review]
Just a couple comments on the hook, will take a look at Zero side changes soon.

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py (hunk 2 above)
Comment (andwgu) [reply torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:58]: Yes. Sorry that I have not had a chance to write the PR text yet explaining the changes. This overlap implementation requires two dummy iterations to warm up, and the training loop still requires `ZeroRedundancyOptimizer.step()` to be called. It becomes a no-op once warmed up.

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py (hunk 4 above)
Comment (andwgu) [reply torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:72]: It happens in [`ZeroRedundancyOptimizer.step()`](https://github.com/pytorch/pytorch/blob/f60106a46aecca09ffa75cb13a179833e6aaff2d/torch/distributed/optim/zero_redundancy_optimizer.py#L831). My understanding: Buckets are rebuilt during the second forward pass; the communication hooks collect the necessary information in the second backward pass; and the state transition happens in the second optimizer step.

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (andwgu) [reply torch/nn/parallel/distributed.py:860]: I see. The idea is that I just need a signal that buckets have been finalized since the design I preferred was to only fully initialize the `ZeroRedundancyOptimizer` instance after that point. This made sense to me given that the current DDP implementation rebuilds buckets early and never again, so any initial construction that seems wasteful, as it will soon be thrown away.

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        the gradients in that bucket.
+
+        Arguments:
+            state: any state for the hook.
+            bucket (dist.GradBucket): the :class:`DistributedDataParallel`
+                gradient bucket.
+        """
+        fut = hook(state, bucket)
+
+        def zero_step(_) -> List[torch.Tensor]:
```
Comment (mrshenli) [inline torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:50]: without get the future argument and calling `Future.wait()` how do we guarantee that ZeRO computation starts after the AllReduce is done?

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py (hunk 2 above)
Comment (mrshenli) [reply torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:58]: (can be done in followup PRs) is it possible to let ZeRO to adapt to DDP's rebuilt buckets?

Comment (mrshenli) [reply torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:58]: Here, we might also need to get the Future as argument, and return `[Future.wait()]`, otherwise `wait()` on the child future might not wait for the allreduce to finish.

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py (hunk 5 above)
Comment (andwgu) [reply torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:50]: I believe that `wait()` is called in the `Reducer` on the future returned by the hook, according to Rohan [here](https://github.com/pytorch/pytorch/pull/61678/files#r674440741). I do not have the code pointer at the moment, but that would explain why we do not need a `wait()` call here. I am not sure if this is accurate since I think that I needed the `wait()` call in the two-stream setting to ensure correctness.

Also, for reference, [Rohan's PR for `hook_then_optim()`](https://github.com/pytorch/pytorch/blob/00f3289d7d4c4ce999a07a6f1e8630fdb1cada38/torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py#L94).

Comment (andwgu) [reply torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:50]: I think perhaps this is also due to the semantics of [`then()`](https://pytorch.org/docs/stable/futures.html#torch.futures.Future.then):

> In other words, as long as the callback doesn’t switch streams, it can safely manipulate the result without any additional synchronization.

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            # properly initialized yet, collect the information needed
+            if overlap_info.status == _OverlapStatus.DDP_HAS_REBUILT_BUCKETS:
+                params_per_rank = overlap_info.params_per_rank
+                params_per_bucket = overlap_info.params_per_bucket
+                if rank_to_update == rank:
+                    overlap_info.offsets[bucket_index] = len(params_per_rank[rank_to_update])
+                params_per_rank[rank_to_update].extend(bucket_params)
+                params_per_bucket.append(bucket_params)
+
+                return [bucket.get_tensor()]
```
Comment (mrshenli) [inline torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:80]: ditto

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                for i, grad in enumerate(bucket_gradients):
+                    gradients[offset + i] = grad
+                zero._local_step(gradients)
+
+            device = bucket_params[0].device
+            device_index = zero._device_to_device_index[device]
+            assert bucket_index in zero._buckets[device_index][rank_to_update]
+            dist.broadcast(zero._buckets[device_index][rank_to_update][bucket_index], src=rank_to_update, async_op=True)
+
+            return [bucket.get_tensor()]
```
Comment (mrshenli) [inline torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:106]: Would I be correct if I assume that since this returned gradients won't be used anyway, we can return arbitrary Tensor values?

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+                offset = overlap_info.offsets[bucket_index]
+                bucket_gradients = bucket.get_per_parameter_tensors()
+                for i, grad in enumerate(bucket_gradients):
+                    gradients[offset + i] = grad
+                zero._local_step(gradients)
+
+            device = bucket_params[0].device
+            device_index = zero._device_to_device_index[device]
+            assert bucket_index in zero._buckets[device_index][rank_to_update]
+            dist.broadcast(zero._buckets[device_index][rank_to_update][bucket_index], src=rank_to_update, async_op=True)
```
Comment (mrshenli) [inline torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:104]: does it make sense to hold on to the returned work handle here and in the last bucket wait for them all. So that when backward returns, all optimizer step comp/comm ops are inserted into the stream or correctly synchronized?

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py (hunk 5 above)
Comment (mrshenli) [reply torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:50]: > I think perhaps this is also due to the semantics of then():

I see. Makes sense, this is enforced here:

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/ivalue_inl.h#L649

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py (hunk 2 above)
Comment (andwgu) [reply torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:58]: Commenting for completeness:

The `Future.wait()` here is not needed if using a single stream for computation since `then()` performs the needed synchronization to guarantee the all-reduce finishes. However, it is needed if the optimizer computation occurs in a separate stream. (Note: The API has changed so that the return should look like `return bucket.get_tensor()` now. However, the current all-reduce hook returns a singleton `List[Tensor]` still, so the call would look like `return fut.wait()[0]`, where `fut` is the argument to `zero_step()`.)

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py (hunk 6 above)
Comment (andwgu) [reply torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:80]: See https://github.com/pytorch/pytorch/pull/62052#discussion_r675670710.

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py (hunk 8 above)
Comment (andwgu) [reply torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:104]: Good catch. I think that this is needed.

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py (hunk 7 above)
Comment (andwgu) [reply torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:106]: I do not believe that assumption is correct. I think that we must still return the bucket's tensor since the value of the new `Future` returned by `then()` will be used in `Reducer::finalize_backward()`. See [here](https://github.com/pytorch/pytorch/blob/04bd9d7577ff855419adca1eb59b58d0f73a1a18/torch/csrc/distributed/c10d/reducer.cpp#L1449).

Comment (rohan-varma) [reply torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py]: Although maybe it shouldn't matter, since the returned gradients are basically just copied back into `param.grad` field, but param.grad field may be irrelevant if we are doing an explicit step on the gradients here.

Although I think you mentioned that before buckets are rebuilt user will still call `zero.step()`, in that case the correct gradient tensor indeed needs to be returned.

**Comment by andwgu** [comment]
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29895933).

**Comment by andwgu** [comment]
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29895933).

**Comment by andwgu** [comment]
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29895933).

**Comment by andwgu** [comment]
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29895933).

**Comment by andwgu** [comment]
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29895933).

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py (hunk 7 above)
Comment (andwgu) [reply torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py]: > Although I think you mentioned that before buckets are rebuilt user will still call zero.step(), in that case the correct gradient tensor indeed needs to be returned.

To clarify this, the `ZeroRedundancyOptimizer.step()` called before buckets are rebuilt do not actually do any optimization, so given what @rohan-varma explained, I think actually we do not need to return the bucket tensor -- just a tensor of the same dimension. I misspoke in the previous comment.

**Comment by andwgu** [comment]
Combined the two approaches; refer to this other PR: https://github.com/pytorch/pytorch/pull/62157
